java/arcs/core/analysis/ExpressionClaimDeducer.kt

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.analysis
+
+import arcs.core.data.Claim
+import arcs.core.data.expression.Expression
+
+/**
+ * A visitor to help create [Claim]s for Paxel [Expression]s.
+ *
+ * For each [Expression], this visitor produces a [Deduction], which can be translated into a set
+ * of [Claim] relationships.
+ *
+ * [Deduction]s collect [Expression.FieldExpression]s as [Deduction.Equal]s and associate
+ * them with other fields. These are recursive structures, and can represent [Claim] relationships
+ * for deeply nested [Expression]s.
+ */
+class ExpressionClaimDeducer : Expression.Visitor<Deduction, Unit> {
+  override fun <E, T> visit(expr: Expression.UnaryExpression<E, T>, ctx: Unit): Deduction {
+    TODO("Not yet implemented")
+  }
+
+  override fun <L, R, T> visit(expr: Expression.BinaryExpression<L, R, T>, ctx: Unit): Deduction {
+    return expr.left.accept(this, ctx) + expr.right.accept(this, ctx)
+  }
+
+  override fun <T> visit(expr: Expression.FieldExpression<T>, ctx: Unit): Deduction {
+    return when (val lhs = expr.qualifier?.accept(this, ctx) ?: Deduction.Equal()) {
+      is Deduction.Equal -> Deduction.Equal(lhs.path + expr.field)
+      is Deduction.Scope -> requireNotNull(lhs.associations[expr.field]) {
+        "Identifier '${expr.field}' is not found in Scope."
+      }
+      is Deduction.Derive -> throw UnsupportedOperationException(
```

Comment:
Having such exceptions indicates that there is a scope for changing the hierarchy a bit or adding some default implementations in the base class.

Reply:
I'm not sure what you mean, could you explain?

Reply:
I don't see how we could restructure things without fundamentally refactoring all of Paxel - and I'm not sure if it's doable even then. From the AST point of view you need to be lookup fields on arbitrary expressions pretty much, e.g. it's ok to lookup fields on results of functions.

Our claim here is that the expression cannot be valid if you're looking up a field on something that produces a `Deduction.Derive` . E.g. `(a + b + 42).x` is not a valid expression.

Soon we will be type checking Paxel expressions (see inflight PR by Ray) and this would be caught at an earlier stage and `UnsupportedOperationException` would not be thrown here.